src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp

```diff
@@ -338,6 +340,764 @@ void dfcc_instrumentt::instrument_function(
     function_pointer_contracts);
 }
 
+static bool is_loop_id_pragma(const irep_idt &pragma)
+{
+  return id2string(pragma).rfind(CONTRACT_PRAGMA_loop_id, 0) == 0;
+}
+
+static size_t get_loop_id_from_pragma(const irep_idt &pragma)
+{
+  INVARIANT(
+    is_loop_id_pragma(pragma), "Wrong usage of get_loop_id_from_pragma");
+  const auto pragma_s = id2string(pragma);
+  size_t last_index = pragma_s.find_last_not_of("0123456789");
+  return atoi(pragma_s.substr(last_index + 1).c_str());
+}
+
+size_t dfcc_instrumentt::get_loop_id_from_target(
+  const goto_programt::instructiont::targett target)
+{
+  INVARIANT(
+    is_loop_instruction(target), "Wrong usage of get_loop_id_from_target");
+  const auto &pragma = get_loop_id_pragma(target);
+
+  return get_loop_id_from_pragma(pragma);
+}
+
+bool dfcc_instrumentt::is_loop_instruction(
+  const goto_programt::instructiont::targett target)
+{
+  const auto &pragma = get_loop_id_pragma(target);
+  return pragma != empty_pragma;
+}
+
+void dfcc_instrumentt::tag_loop_instruction(
+  goto_programt::instructiont::targett &target,
+  const irep_idt &pragma,
+  const size_t loop_id)
+{
+  target->source_location_nonconst().add_pragma(
+    id2string(pragma) + ":" + std::to_string(loop_id));
+}
+
+irep_idt
+dfcc_instrumentt::get_loop_id_pragma(const goto_programt::targett &target)
+{
+  for(const auto &pragma : target->source_location().get_pragmas())
+  {
+    if(is_loop_id_pragma(pragma.first))
+      return pragma.first;
+  }
+  return empty_pragma;
+}
+
+void dfcc_instrumentt::remove_loop_id_pragma(goto_programt::targett &target)
+{
+  for(const auto &pragma : target->source_location().get_pragmas())
+  {
+    if(is_loop_id_pragma(pragma.first))
+    {
+      target->source_location_nonconst()
+        .get_named_sub()
+        .add(ID_pragma)
+        ->second.remove(pragma.first);
+      return;
+    }
+  }
+}
```

Comment:
Could we avoid looping over all pragmas and having to parse the string each time ?

How about using something like this:

```
const irep_idt ID_loop_id = "loop-id";
const irep_idt ID_loop_instr_type = "loop-instr-type";
const irep_idt ID_loop_instr_body = "body";
const irep_idt ID_loop_instr_head = "head";
const irep_idt ID_loop_instr_latch = "latch";
const irep_idt ID_loop_instr_exiting = "exiting";

void tag_loop_instr(targett &target, const irep_idt &instr_type, const std::size_t loop_id)
{
    irept &irep = target->source_location_nonconst();
    irep.set(ID_loop_instr_type, instr_type);
    irep.set(ID_loop_id, loop_id);
}

irep_idt &get_loop_instr(targett &target)
{
    irept &irep = target->source_location();
    return irep.get(ID_loop_instr_type);
}

std::size_t get_loop_id(targett &target)
{
    PRECONDITION(!get_loop_instr_type(target).empty());
    irept &irep = target->source_location();
    return irep.get_size_t(ID_loop_id);
}

bool is_loop_instr(targett &target, std::size_t loop_id) {
    if (get_loop_instr_type(target).empty()) return false;
    return get_loop_id(target) == loop_id;
}
```